### PR TITLE
Switch back envoy upstream to envoy-wasm

### PR DESCRIPTION
**Pull request by bianpengyuan**
Also use bazel to build Wasm modules. Fix https://github.com/istio/istio/issues/23215

### CODE REVIEW

File: Makefile.core.mk [hunk 1]
```diff
@@ -76,21 +76,12 @@ build_envoy_tsan:
...
-	cp -f $$(bazel info output_base)/external/envoy/api/wasm/cpp/contrib/proxy_expr.h $(TOP)/extensions/common/
-
-build_wasm: wasm_include
-	$(foreach file, $(shell find extensions -name build_wasm.sh), cd $(TOP)/$(shell dirname $(file)) && bash ./build_wasm.sh &&) true
-
-check_wasm:
-	export PATH=$(PATH) CC=$(CC) CXX=$(CXX) && bazel $(BAZEL_STARTUP_ARGS) build $(BAZEL_BUILD_ARGS) $(BAZEL_CONFIG_DEV) //src/envoy:envoy
-	./scripts/generate-wasm.sh -b
+build_wasm:
+	bazel build //extensions/stats:plugin.wasm
```
Comment (mandarjog) [inline Makefile.core.mk:80]: if you want to add this to the make file, we should use `BAZEL_STARTUP_ARGS`, `CC`, `CXX` all standard variable.

File: bazel/patches/absl.patch [hunk 2]
```diff
@@ -0,0 +1,22 @@
+diff --git a/absl/base/BUILD.bazel b/absl/base/BUILD.bazel
```
Comment (bianpengyuan) [inline bazel/patches/absl.patch:1]: I cannot find a better way to work around this. Inside google we have:
```
        "//tools/cc_target_os:windows": [],
        # absl:google3-begin(emscripten is google3 only)
        "//tools/cc_target_os:emscripten": [],
        # absl:google3-end
        "//conditions:default": ["-pthread"],
```

We probably need to add the same options to upstream absl. @lizan @jplevyak @PiotrSikora Any suggestion?

File: Makefile.core.mk (hunk 1 above)
Comment (bianpengyuan) [reply Makefile.core.mk:80]: But those ENV variable, or build options do not really apply to the Wasm target, right?

**Comment by bianpengyuan** [comment]
rbe fails to build wasm module due to missing std library header:
```

Use --sandbox_debug to see verbose messages from the sandbox
In file included from external/com_github_google_flatbuffers/src/reflection.cpp:17:
In file included from bazel-out/host/bin/external/com_github_google_flatbuffers/src/_virtual_includes/flatbuffers/flatbuffers/reflection.h:25:
In file included from bazel-out/host/bin/external/com_github_google_flatbuffers/src/_virtual_includes/flatbuffers/flatbuffers/reflection_generated.h:7:
In file included from bazel-out/host/bin/external/com_github_google_flatbuffers/src/_virtual_includes/flatbuffers/flatbuffers/flatbuffers.h:20:
bazel-out/host/bin/external/com_github_google_flatbuffers/src/_virtual_includes/flatbuffers/flatbuffers/base.h:28:10: fatal error: 'cstdint' file not found
#include <cstdint>
         ^~~~~~~~~
1 error generated.
Target //extensions:stats.wasm failed to build
Use --verbose_failures to see the command lines of failed build steps.
ERROR: /home/prow/go/src/istio.io/proxy/extensions/BUILD:29:1 C++ compilation of rule '@com_github_google_flatbuffers//grpc/src/compiler:swift_generator' failed (Exit 1) clang failed: error executing command /usr/lib/llvm/bin/clang -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 24 argument(s) skipped)

```

File: bazel/patches/absl.patch (hunk 2 above)
Comment (bianpengyuan) [reply bazel/patches/absl.patch:1]: Created https://github.com/abseil/abseil-cpp/issues/693 to track this.

Comment (lizan) [reply bazel/patches/absl.patch:1]: Yeah we definitely need a right select there, but currently bazel doesn't define a common cpu/platform for wasm. I raised PR https://github.com/bazelbuild/platforms/pull/12 but not merged yet.

**Comment by lizan** [comment]
cc @Shikugawa

File: extensions/stackdriver/common/utils.h [hunk 3]
```diff
@@ -16,11 +16,17 @@
...
-#include "envoy/config/core/v3/grpc_service.pb.h"
 #include "extensions/common/context.h"
 #include "google/api/monitored_resource.pb.h"
 #include "grpcpp/grpcpp.h"
 
+#ifndef NULL_PLUGIN
+#include "api/wasm/cpp/proxy_wasm_intrinsics.h"
+#else
+#include "envoy/config/core/v3/grpc_service.pb.h"
+using GrpcService = ::envoy::config::core::v3::GrpcService;
```
Comment (gargnupur) [inline extensions/stackdriver/common/utils.h:27]: nit: avoid "using" in header file?

**Pull request merged**